server/tests/test_pipeline.py

```diff
@@ -82,3 +83,105 @@ def test_to_dict():
+            ]
+        },
+    }
+
+    composed_filter_step_or_and_ = {
+        "name": "filter",
+        "condition": {
+            "or_": [
+                {"column": "colE", "operator": "eq", "value": "{{ __front_var_3__ }}"},
+                {"column": "colF", "operator": "lt", "value": "{{ __front_var_4__ }}"},
+                {
+                    "and_": [
+                        {"column": "colG", "operator": "gt", "value": "{{ __front_var_5__ }}"},
+                        {"column": "colH", "operator": "eq", "value": "{{ __front_var_6__ }}"},
+                        {
+                            "or_": [
+                                {"column": "colI", "operator": "eq", "value": "toto"},
+                                {
+                                    "column": "colJ",
+                                    "operator": "eq",
+                                    "value": "{{ __front_var_7__ }}",
+                                },
+                            ]
+                        },
+                    ]
+                },
+            ]
+        },
+    }
+
+    steps = [
+        {"domain": "foobar", "name": "domain"},
+        void_step,
+        composed_filter_step_or_and_,
+        simple_filter_step,
+        composed_filter_step_and_,
+    ]
+    variables = {
+        "__front_var_0__": "__VOID__",
+        "__front_var_1__": 32,
+        "__front_var_2__": "__VOID__",
+        "__front_var_3__": "TEST TEST",
+        "__front_var_4__": "__VOID__",
+        "__front_var_5__": "__VOID__",
+        "__front_var_6__": "VALUE",
+        "__front_var_7__": "TOTO",
+    }
+
+    pipeline_with_variables = PipelineWithVariables(steps=steps)
+    pipeline = pipeline_with_variables.render(variables, renderer=_render_query)
+
+    # will render the with no __VOID__ in step, whatever the depth
+    assert [s.dict() for s in pipeline.steps] == [
+        {"domain": "foobar", "name": "domain"},
+        {
+            "condition": {
+                "or_": [
+                    {"column": "colE", "operator": "eq", "value": "TEST TEST"},
+                    {
+                        "and_": [
+                            {"column": "colH", "operator": "eq", "value": "VALUE"},
+                            {
+                                "or_": [
+                                    {"column": "colI", "operator": "eq", "value": "toto"},
+                                    {"column": "colJ", "operator": "eq", "value": "TOTO"},
+                                ]
+                            },
+                        ]
+                    },
+                ]
+            },
+            "name": "filter",
+        },
+        {"condition": {"column": "colB", "operator": "eq", "value": 32}, "name": "filter"},
+        {
+            "condition": {
+                "and_": [
+                    {"column": "ColD", "operator": "until", "value": datetime(2009, 1, 3, 0, 0)}
+                ]
+            },
```

Comment:
un `and_` avec une liste d'un seul élément devrait être simplifié en enlevant le `and_` IMO (à moins que nos différents traducteurs/exécuteurs gèrent ce cas)
:information_source: Il faut aussi penser au cas où on se retrouve avec 0 éléments and un and_ ou un or_

Reply:
> Il faut aussi penser au cas où on se retrouve avec 0 éléments and un and_ ou un or_

good catch, handled here : [`be77ded` (#1763)](https://github.com/ToucanToco/weaverbird/pull/1763/commits/be77dedb2e13939d52ccd6b8f314d4954510f575)